Making Scheduler ready for MultiServer

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->
 
#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
Make Scheduler ready for Multi-Server. Following are the major things that are done to achieve this goal.

- Scheduler maintaining persistent server connections
- Change of method to notify Server/s for end of cycle notification as we cannot close connections for every scheduling cycle
- Making Scheduler _select IO_ framework ready to accept/process requests from Multiple Servers
- Libshard integration with Scheduler
#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->

- Scheduler now accepts and process requests from Multiple Servers
- Scheduler maintains persistent server connections
- As the connections are persistent, when Scheduler finishes its cycle it needs to have a mechanism through which it can tell the Server that scheduling cycle is finished. This can be easily achieved by just sending a flag on secondary socket which acts as an end of cycle indication to the Server.
- Each Server can mark a flag(part of pbs_sched structure) when it kicks off a scheduling cycle and clears this flag when it receives end of cycle indication from Scheduler.
- Server should receive end of cycle notification in process_request function when sched cycle finishes.  In this function we need to check every time whether each request is from Scheduler or not which is a O(n) operation(if there are Multiple Schedulers). Optimised this operation to O(1) by introducing is_sched_conn in connection structure and the corresponding code changes.
- **How Multiple Servers distinguish themselves to Scheduler**
Each Server has a unique identifier whose value can be obtained by using get_svr_index(). A Server sends its identity to Scheduler via a new command SCH_SVR_IDENTIFIER at the time of connection initiation. As the connections are persistent Server does not send this identifier thereafter.
- Re-factor Server side code changes so as to work with the above changes.
- **Why are TCP_USER_TIMEOUT and SO_REUSEPORT required ?**
- SO_REUSEPORT is to avoid EADDRINUSE. We are already using SO_REUSEADDR but I learnt that this is not enough in some systems and hence the option SO_REUSEPORT is also needed.
- At times Server doesn't receive a notification as soon as Scheduler is either shutdown or abruptly killed. It might take minutes for the Server to know that Scheduler is not active. Also there are lot of test cases(especially Multisched) which validate Server's state as soon as Scheduler is shutdown. If the Server does not immediately know that Scheduler is shutdown sched object's state is different and lot of test cases will fail.
- **Libshard integration with Scheduler**
As some of the large scale objects like jobs in PBS are sharded in case of Multi-Server, Scheduler should be able to able to talk to to the right Server. Each IFL called from Scheduler picks up the right server by calling get_svr_shard_connection with the help of the parameters socket(virtual or real in case of single server), object type and object id. This function internally calls Libshard APIs to fetch the right server which mandates that Scheduler is integrated with Libshard. Following are high level changes to achieve the same.

- Call pbs_shard_init() to initialise Libshard.
- In case of external clients like qsub, qstat etc. ch_shards table is created and auto populated by get_svr_shard_connection() function by calling the appropriate connect function which connects to the Server. It uses these table thereafter. When it comes to Scheduler since Server only connects to Scheuuler,  this table should not be created by get_svr_shard_connection. Rather by the time get_svr_shard_connection() gets called ch_shards table should be created and should be handy. This is the key in integration. get_svr_shard_connection() is modified accordingly.
- So Scheduler initialises ch_shards table based on the number of currently configured servers in the beginning. Please look into the function initialise_svr_sock_context() for more details.
- Scheduler fills up ch_shards table as and when a Server connects to the Scheduler. Please look at the functions accept_client() and socket_to_conn() for more details. Same table can be used to get and process requests from different Servers.
- ch_shards table is ordered as per the Server id or index. This is one of the requirements from Libshard to call its respective APIs.

**Miscellaneous changes:**
- At times Server can crash if there is no call back registered when a connection gets closed.
- No need to validate certain Scheduler attributes during recovery from db.




#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
Following is the design document.
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1518108673/Scheduler+Maintaining+Persistent+Server+Connections

#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->

Following are TH3 regression logs.
<img width="451" alt="image" src="https://user-images.githubusercontent.com/17923736/81048651-59461000-8eda-11ea-82b4-7b2505060369.png">

**How Multi-Server code path is tested**
As we cannot test complete Multi-Server functionality at the moment I have done the following Unit testing to make sure that Multi-Server code path hits and works fine.

• Configure PBS_MAX_SERVERS to 3 and PBS_SERVER_INSTANCES to contain only one active server. This is the only configuration we can test at the moment.r

    root@stblr3 pbspro]# cat /etc/pbs.conf
    PBS_SERVER=stblr3
    PBS_START_SERVER=1
    PBS_START_SCHED=1
    PBS_START_COMM=1
    PBS_START_MOM=1
    PBS_EXEC=/opt/pbs
    PBS_HOME=/var/spool/pbs
    PBS_CORE_LIMIT=unlimited
    PBS_SCP=/bin/scp
    PBS_MAX_SERVERS=3
    PBS_SERVER_INSTANCES=stblr3:15001
• The above configuration also does not work at the moement since job sequence id generation code changes are not yet present. In other words each server is generating job ids by not following any sharding logic.
• In the event of the above,  I had to tweak pbs_shard_get_server_byindex API to make sure that it returns the index of configured active server which is only one as per the above configuration. By doing this it ensures that we can unit test the multi-server code path. Following are the details of this testing.

[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
40.stblr3
41.stblr3
42.stblr3
43.stblr3
44.stblr3
45.stblr3
46.stblr3
47.stblr3
48.stblr3
49.stblr3


[root@stblr3 pbspro]# qstat -s
[root@stblr3 pbspro]# 

**sched_logs snippet:**
[root@stblr3 sched_logs]# tail -f 20200504 
05/04/2020 10:25:09;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:25:09;0080;pbs_sched;Job;39.stblr3;Considering job to run
05/04/2020 10:25:09;0040;pbs_sched;Job;39.stblr3;Job run
05/04/2020 10:25:09;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:25:09;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:25:09;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:25:09;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:25:09;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:25:09;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:25:09;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;40.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;40.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;41.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;41.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;42.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;42.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;43.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;43.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;44.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;44.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;45.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;45.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;46.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;46.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;47.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;47.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;48.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;48.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/04/2020 10:27:01;0080;pbs_sched;Job;49.stblr3;Considering job to run
05/04/2020 10:27:01;0040;pbs_sched;Job;49.stblr3;Job run
05/04/2020 10:27:01;0080;pbs_sched;Req;;Leaving Scheduling Cycle

**Shutdown Server and then start it, submit a few jobs, make sure all of them run**
[root@stblr3 pbspro]# qterm
[root@stblr3 pbspro]# service pbs status
pbs_server is not running
pbs_mom is pid 105200
pbs_sched is pid 105213
pbs_comm is 105169
[root@stblr3 pbspro]# pbs_server
Connecting to PBS dataservice....connected to PBS dataservice@stblr3
Licenses valid for 10000000 Floating hosts
[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
60.stblr3
61.stblr3
62.stblr3
63.stblr3
64.stblr3
65.stblr3
66.stblr3
67.stblr3
68.stblr3
69.stblr3
[root@stblr3 pbspro]# qstat -s

**Testing of qterm -s(which terminates both server and scheduler)**
[root@stblr3 pbspro]# qterm -s
[root@stblr3 pbspro]# service pbs status
pbs_server is not running
pbs_mom is pid 105200
pbs_sched is not running
pbs_comm is 105169

**Test submitting jobs after the above scenario**
[root@stblr3 pbspro]# pbs_sched;pbs_server;
Connecting to PBS dataservice....connected to PBS dataservice@stblr3
Licenses valid for 10000000 Floating hosts
[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
70.stblr3
71.stblr3
72.stblr3
73.stblr3
74.stblr3
75.stblr3
76.stblr3
77.stblr3
78.stblr3
79.stblr3
[root@stblr3 pbspro]# qstat -s

**Shutdown scheduler alone, bring it back and test submitting jobs**
[root@stblr3 pbspro]# pkill pbs_sched
[root@stblr3 pbspro]# service pbs status
pbs_server is pid 109511
pbs_mom is pid 105200
pbs_sched is not running
pbs_comm is 105169
[root@stblr3 pbspro]# pbs_sched
[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
80.stblr3
81.stblr3
82.stblr3
83.stblr3
84.stblr3
85.stblr3
86.stblr3
87.stblr3
88.stblr3
89.stblr3
[root@stblr3 pbspro]# qstat -s

**MultiServer code path testing after rebase**
[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
0.stblr3
3.stblr3
6.stblr3
9.stblr3
12.stblr3
15.stblr3
18.stblr3
21.stblr3
24.stblr3
27.stblr3
[root@stblr3 pbspro]# qstat -s
[root@stblr3 pbspro]# 


**sched_logs snippet:**
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;0.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;0.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;3.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;3.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;6.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;6.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;9.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;9.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;12.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;12.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;15.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;15.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;18.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;18.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;21.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;21.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;24.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;24.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Req;;Starting Scheduling Cycle
05/12/2020 12:09:28;0080;pbs_sched;Job;27.stblr3;Considering job to run
05/12/2020 12:09:28;0040;pbs_sched;Job;27.stblr3;Job run
05/12/2020 12:09:28;0080;pbs_sched;Req;;Leaving Scheduling Cycle



**Shutdown Server and then start it, submit a few jobs, make sure all of them run**
[root@stblr3 pbspro]# qterm
[root@stblr3 pbspro]# service pbs status
pbs_server (on port 15001) is not running
pbs_mom is pid 71110
pbs_sched is pid 71123
pbs_comm is 71091
[root@stblr3 pbspro]# pbs_server
Connecting to PBS dataservice...connected to PBS dataservice@stblr3
Licenses valid for 10000000 Floating hosts
[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
30.stblr3
33.stblr3
36.stblr3
39.stblr3
42.stblr3
45.stblr3
48.stblr3
51.stblr3
54.stblr3
57.stblr3
[root@stblr3 pbspro]# qstat -s
[root@stblr3 pbspro]# 


**Testing of qterm -s(which terminates both server and scheduler)**
[root@stblr3 pbspro]# qterm -s
[root@stblr3 pbspro]# service pbs status
pbs_server (on port 15001) is not running
pbs_mom is pid 71110
pbs_sched is not running
pbs_comm is 71091


**Test submitting jobs after the above scenario**
[root@stblr3 pbspro]# pbs_sched;pbs_server;
Connecting to PBS dataservice...connected to PBS dataservice@stblr3
Licenses valid for 10000000 Floating hosts
[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
60.stblr3
63.stblr3
66.stblr3
69.stblr3
72.stblr3
75.stblr3
78.stblr3
81.stblr3
84.stblr3
87.stblr3
[root@stblr3 pbspro]# qstat -s
[root@stblr3 pbspro]# 

**Shutdown scheduler alone, bring it back and test submitting jobs**
[root@stblr3 pbspro]# pkill pbs_sched
[root@stblr3 pbspro]# service pbs status
pbs_server (on port 15001) is pid 72354
pbs_mom is pid 71110
pbs_sched is not running
pbs_comm is 71091
[root@stblr3 pbspro]# pbs_sched
[root@stblr3 pbspro]# for i in $(seq 1 10); do qsub -koe -- date; done
90.stblr3
93.stblr3
96.stblr3
99.stblr3
102.stblr3
105.stblr3
108.stblr3
111.stblr3
114.stblr3
117.stblr3
[root@stblr3 pbspro]# qstat -s
[root@stblr3 pbspro]# 
<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

suresh-thelkar commented:
Closed the following PR as the current PR is superset which also contains Scheduler integration with Libshard. Also adding the link to the previous PR for reference.
https://github.com/subhasisb/pbspro/pull/73

subhasisb commented:
@suresh-thelkar  your PR shows 163 file differences. That is absurd. Can you please submit only your changes related to the scheduler ready for multiserver?

agrawalravi90 commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
+					log_eventf(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__,
+							"get_sched_cmd failed, errno=%d in function %s. "
+							"One  of the reasons includes server might have shutdown", errno, __func__);
+				} else {
```
This is a big else, again I'd suggest you to put this into a separate method

suresh-thelkar replied:
This block of code is also moved to the new function schedule_wrapper.

agrawalravi90 commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
+						 */
+						sock_schedule = server_sock;
+					} else {
+						sock_schedule = sock_to_check;
```
even for single server case, it is server_sock right? I think this if-else block is not  needed

suresh-thelkar replied:
No. We are supposed to use server_sock only for Multi-Sever. We consider this as a virtual socket.
For single server we use sock_to_check.

agrawalravi90 commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
+						sock_schedule = sock_to_check;
+					}
+
+					if (sock_to_check >= 0) {
```
if sock_to_check was less than 0 then you wouldn't have entered the "if ((sock_to_check != -1) && FD_ISSET(sock_to_check, &read_fdset)) " block above, so this if condition is not needed

suresh-thelkar replied:
Implemented.

agrawalravi90 commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
+	#endif /* localmod 031 */
+
+					sched_ret = schedule(cmd, sock_schedule, runjobid);
+					if (sched_ret != 0 )  /* magic happens here */ {
```
the comment seems misplaced

suresh-thelkar replied:
corrected the placement of the comment.

agrawalravi90 commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
+						close_server_conns(index_to_shards);
+						if (cmd == SCH_QUIT) {
+							go = 0;
+							break;
```
do you need to call sigprocmask before breaking out?

suresh-thelkar replied:
Yes. I have added it in the new commit.

agrawalravi90 commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1547,108 @@ main(int argc, char *argv[])
+                }
+                return set_conn_shards_exist(server_sock);
+	} else {
+		svr_sock_pair.ch_socket= -1;
```
why not just use the same data structure as  multi-server mode ? it seems like there's no operational difference between the 2 modes for sched, it's right now just a difference of the data structure used, which seems unnecessary to me. Making them the same will clean up so much of the code, remove so many conditionals.

suresh-thelkar replied:
As we discussed the next goal is to make sure we have same code path for both single and Multi-Server. Yes we are going to reach there very shortly. 

But unfortunately we cannot do this in this PR at the moment due to the following reasons.

- ch_shards table exists only for the Multi-Server case in our integration branch.
- Libshard APIs, get_svr_shard_connection and all IFLs which use are coded assuming the above fact.
- Lot of PTL test cases need to modified if we now use ch_shards table for single server also.

So we are planning to reach to our next goal after this PR and MOM integration with Libshard PR are merged.

agrawalravi90 replied:
Ok .. just curious, why will PTL tests need to be modified?

suresh-thelkar replied:
Some new PBS_* error code are introduced as part of Libshard PR which works only for Multi-Server. So if we go with the same code path then we need to change PTL test cases to handle the same. You can refer the following discussion for more details.
https://github.com/subhasisb/pbspro/pull/81#discussion_r422749532

suresh-thelkar commented:
> @suresh-thelkar your PR shows 163 file differences. That is absurd. Can you please submit only your changes related to the scheduler ready for multiserver?

@subhasisb, It looks like this happened because integration branch got rebased with master very recently. I will rebase my branch with the integration branch, do some testing and let you know after that.

suresh-thelkar commented:
@subhasisb / @agrawalravi90, I have rebased it with the integration branch. Also updated the conversation section with the latest testing done for MultiServer code path. I will continue to do further testing.

PR is ready for further review. Please provide your comments.

subhasisb commented:
Why do you need SO_REUSEPORT? That seemed to have been introduced to the kernel to handle multiple threads doing accept situations - see https://lwn.net/Articles/542629/

suresh-thelkar commented:
> https://lwn.net/Articles/542629/

Thanks for sharing the link. Following are the reasons why I have used SO_REUSEPORT.

- I have observed couple of Multisched test cases were failing intermittently with the error "Address already in use" as soon the Multischeds were restarted. There are some test cases in Multisched which abruptly kills the scheduler/s(without a graceful shutdown) and are brought back during which case the issue was happening. 
- An observation is that I have never seen the issue when Scheduler was not multithreaded. 
- In order to resolve the issue I have used SO_RESUSEPORT after which the issue did not happened forever.

It looks like from the link you shared SO_REUSEPORT is safer than SO_REUSEADDR because it will check the effective user id of the process and since Scheduler is multithreaded I feel it is ok to use the option. I might be wrong though. Please correct me if my understanding is wrong.

Please suggest if I have to revert back the option.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -563,86 +640,40 @@ server_command(char **jid)
-		CS_close_socket(new_socket);
+       if (cmd == SCH_SVR_IDENTIFIER)
+		i = strtol(svr_id, &endp, 10);
+	else {
```
No. We don't have a job id when the control reaches this place. As this function gets called only for new connection we always get SCH_SVR_IDENTIFIER.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -563,86 +640,40 @@ server_command(char **jid)
-		log_err(errno, __func__, "get_sched_cmd");
-		CS_close_socket(new_socket);
+       if (cmd == SCH_SVR_IDENTIFIER)
+		i = strtol(svr_id, &endp, 10);
```
Have renamed i to svr_index.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -563,86 +640,40 @@ server_command(char **jid)
-		log_err(errno, __func__, "get_sched_cmd");
-		CS_close_socket(new_socket);
+       if (cmd == SCH_SVR_IDENTIFIER)
+		i = strtol(svr_id, &endp, 10);
```
Added the validation check.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1547,108 @@ main(int argc, char *argv[])
+static int
+socket_to_conn(int sock, int svr_id)
+{
+
```
Deleted the blank line.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1547,108 @@ main(int argc, char *argv[])
+{
+
+	if (get_max_servers() > 1) {
+		shard_conn_t ** shard_conn;
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1547,108 @@ main(int argc, char *argv[])
+static int
+initialise_svr_sock_context()
+{
+	if (get_max_servers() > 1) {
```
Implemented.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1099,6 +1137,14 @@ main(int argc, char *argv[])
+#ifdef SO_RESUSEPORT
+	if (setsockopt(server_sock, SOL_SOCKET, SO_REUSEPORT,
+		(char *)&t, sizeof(t)) == -1) {
+		log_err(errno, __func__, "setsockopt");
```
I think errno is sufficient to find out the root cause of why setsockopt failed. There are other system calls in the same file(like gethostname, socket etc.) for which only errno is added and I followed the same. If needed to change we have to change all other system calls.

hirenvadalia replied:
@agrawalravi90 Don't worry, log_err will convert errno to string before logging.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1127,7 +1173,7 @@ main(int argc, char *argv[])
 	}
 
-	if (listen(server_sock, 5) < 0) {
+	if (listen(server_sock, LISTEN_BACKLOG) < 0) {
```
We actually need to add all servers to the authorised clients list. Thanks for catching this. I have now added the code for it.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1327,18 +1373,38 @@ main(int argc, char *argv[])
+		fprintf(stderr, "pbs_shard_init failed\n");
+		return -1;
+	} else
+		shard_init_done = 1;
```
Scheduler should initialise Libshard by using the routine pbs_shard_init() only once. As soon as we initialise we set a flag i.e. shard_init_done = 1. This way pbs_shard_init() is not called again when get_svr_shard_connection() is called several times through various IFLs used by Scheduler.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -111,10 +111,13 @@
-int		connector;
 int		server_sock;
 int		second_connection = -1;
+fd_set		master_fdset;
```
select system call modifies the fd_sets in this case read_fdset with the only socket fds that are ready for IO for each iteration.  So after select call read_fdset may not contain all the fds that we are interested in.

To overcome this we keep server_sock and all other new sockets added to master_fdset. This way we can always derive read_fdset for every iteration of select using master_fdset.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1547,108 @@ main(int argc, char *argv[])
+
+		if (shard_conn[svr_id]->sd == -1) {
+			shard_conn[svr_id]->sd = sock;
+			FD_SET(sock, &master_fdset);
```
Please look into my earlier answer for this as well.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
+		if (FD_ISSET(server_sock, &read_fdset)) {
+			if (accept_client(&max_sd) != 0)
+				die(0);
+		}
```
No. We shouldn't have a continue here. Apart from the new connections we also might have old or existing connections by the time control reaches here.

Let us say we have Multi-Server setup and first server already connected to the Scheduler. When second server tries to connect to the Scheduler, here we are calling accept_client(), but by this time IO for Server-1 might be ready and hence we should go to the code after this block.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
-				strftime(log_buffer, sizeof(log_buffer),
-					"%Y-%m-%d %H:%M:%S", localtime(&now));
-				printf("%s Scheduler received command %d\n", log_buffer, cmd);
+		for (index_to_shards = 0; index_to_shards < num_cfg_svrs; index_to_shards++) {
```
Implemented.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1416,127 @@ main(int argc, char *argv[])
-				close(second_connection);
-				second_connection = -1;
+			if ((sock_to_check != -1) && FD_ISSET(sock_to_check, &read_fdset)) {
+				int ret;
```
Have come up with a new function schedule_wrapper() which calls schedule internally and this new function takes care of both single and multi-server.

suresh-thelkar commented:
@agrawalravi90, I have addressed your comments. Please take a look at the latest commit and provide any further feedback.

agrawalravi90 commented on src/include/pbs_sched.h:
```diff
@@ -114,12 +119,21 @@ typedef struct pbs_sched {
+ * connection(either primary or secondary) it has to talk to.
+ */
+enum towhich_conn {
+	PRIMARY,
```
there's another data type called PRIMARY, can we rename these to PRIMARY_CONN & SECONDARY_CONN ?

suresh-thelkar replied:
Done

agrawalravi90 commented on src/include/pbs_sched.h:
```diff
@@ -66,6 +66,11 @@ extern "C" {
 
 #define SC_STATUS_LEN 	10
 
+#define	SCHED_CYCLE_END 0
```
inconsistent indent

suresh-thelkar replied:
It looks like something wrong with GitHub. I do not see any indentation issue in vi editor.

agrawalravi90 commented on src/include/pbs_sched.h:
```diff
@@ -66,6 +66,11 @@ extern "C" {
 
+#define	SCHED_CYCLE_END 0
+#define LISTEN_BACKLOG 100
+#define TCP_TIMEOUT 50000
```
That's kind of huge (~14 hours), do we really need such a big idle timeout value?

suresh-thelkar replied:
This is in milliseconds, so its value is 50 seconds not ~14 hours.

agrawalravi90 commented on src/include/pbs_sched.h:
```diff
@@ -66,6 +66,11 @@ extern "C" {
+#define	SCHED_CYCLE_END 0
+#define LISTEN_BACKLOG 100
+#define TCP_TIMEOUT 50000
+#define SVR_MAX_INDEX 3
```
Please add a comment here to explain what this does

agrawalravi90 replied:
Why only 3? Is this meant to be a default value for MAX_SERVER_INSTANCES?

suresh-thelkar replied:
I have changed the macro value to 2 now which is sufficient and added a comment to describe its use. Please look into my next commit.

agrawalravi90 commented on src/include/libpbs.h:
```diff
@@ -130,6 +130,7 @@ typedef struct pbs_conn {
 	char *ch_errtxt;                /* pointer to last server error text	*/
 	void *ch_shards;                /* handles to multiple sharded servers */
 	int shard_context;              /* last server to which dialogue was going on */
+	int ch_shards_exist;		/* used to know whether to create ch_shards or not */
```
ch_shards == NULL should tell us if it exists or not right?

suresh-thelkar replied:
It is not enough to just know whether ch_shards exists or not. We should actually know whether this table is filled by a client(for example by scheduler or not). I have renamed a few variables here to make it more readable. Please look into my next commit for more details.

agrawalravi90 commented on src/lib/Libifl/conn_table.c:
```diff
@@ -662,3 +662,74 @@ get_conn_shards(int vfd)
+ * @par MT-safe: Yes
+ */
+int
+get_conn_shards_exist(int vfd)
```
I think what you really want from this function is to know whether ch_shards is set or not. So, why not just name it something like "is_conn_shards_set" or "does_conn_shards_exist" ?

suresh-thelkar replied:
I have renamed the function to is_conn_shards_filled(). The idea is we really need to know whether ch_shards table is filled up by a client or not.

agrawalravi90 commented on src/include/pbs_sched.h:
```diff
@@ -114,12 +119,21 @@ typedef struct pbs_sched {
+/* This is used when connecting to the Scheduler to let Server know on which
+ * connection(either primary or secondary) it has to talk to.
+ */
+enum towhich_conn {
```
how about just "sched_conn_t" since this is only used for talking to sched?

suresh-thelkar replied:
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -504,33 +576,34 @@ badconn(char *msg)
  */
 int
-server_command(char **jid)
+accept_client(int *max_sd)
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -563,86 +643,46 @@ server_command(char **jid)
+	       }
+       }
+	else {
+		log_err(errno, __func__, "Invalid Server id");
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -563,86 +643,46 @@ server_command(char **jid)
-			second_connection = -1;
-			return cmd;
-		}
+       if (svr_id != NULL) {
```
Removed the NULL check.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -837,6 +876,10 @@ main(int argc, char *argv[])
 	int update_svr = 1;
+	int num_cfg_svrs;
+	int max_svrs;
+	shard_conn_t **shard_conn = NULL;
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -837,6 +876,10 @@ main(int argc, char *argv[])
+	int num_cfg_svrs;
+	int max_svrs;
+	shard_conn_t **shard_conn = NULL;
+	int shard_inst_idx;
```
Done

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -662,3 +662,74 @@ get_conn_shards(int vfd)
+	if (INVALID_SOCK(vfd))
+		return -1;
+
+	LOCK_TABLE(-1);
```
Not required as LOCK_TABLE returns -1 in case of failure. Instead added a check at the caller of this statement.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -455,34 +483,61 @@ int
+		ch_shards_exist = get_conn_shards_exist(vsock);
+	}
+
+	if (ch_shards_exist) {
```
I have renamed the variable to ch_shards_filled for better readability and also added comments.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -522,12 +577,12 @@ get_svr_shard_connection(int vsock, enum pbs_obj_type obj_type, void *obj_id)
 					if (!inact_svr_indexes) {
 						inact_servers = 0;
-						if (!(inact_svr_indexes = (int *)malloc(num_of_conf_servers * sizeof(int)))) {
+						inact_svr_indexes = initialise_inact_svr_indexes();
```
The idea is to initialise the array with -1 not with 0 and hence cannot use calloc. Also the same block of code is getting used at two different places in the same function. Due to this I have written a initialiser function.

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -1014,7 +1015,8 @@ net_close(int but)
 		cp = GET_NEXT(cp->cn_link);
 		if(sock != but) {
-			svr_conn[sock]->cn_oncl = NULL;
+			if (svr_conn[sock]->cn_oncl != NULL)
```
I saw server getting crashed with the absence of the above check once and hence added it.

suresh-thelkar replied:
If it was directly set to NULL I observed a crash once. Think of a following case. What if a particular connection hasn't have a function registered during its closure and also not set cn_oncl to NULL. That time we can get a crash and hence added the above check. Also if you look at other functions like close_conn() we are doing the similar thing.

agrawalravi90 replied:
Still not getting you. You are setting cn_oncl to NULL if it's not NULL. I don't understand how the if statement will prevent a crash since after this if block, cn_oncl is guaranteed to be NULL anyways. Inside close_conn() we are doing "svr_conn[idx]->cn_oncl(sd);" if svr_conn[idx]->cn_oncl !=NULL, that's different than setting it to NULL right?

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1470,235 @@ main(int argc, char *argv[])
+					 */
+					sock_schedule = server_sock;
+				} else {
+					sock_schedule = sock_to_check;
```
No server_sock is only for Multi-Server.

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -1593,7 +1593,7 @@ main(int argc, char **argv)
 			/* Bring up scheduler here */
 			pbs_scheduler_addr = get_hostaddr(pbs_conf.pbs_secondary);
-			if (contact_sched(SCH_SCHEDULE_NULL, NULL, pbs_scheduler_addr, pbs_scheduler_port) < 0) {
+			if (contact_sched(SCH_SCHEDULE_NULL, NULL, dflt_scheduler, PRIMARY) < 0) {
```
Because I have changed the prototype of contact_sched. It now accepts pbs_sched structure pointer which actually contains pbs_scheduler_addr and pbs_scheduler_port which are actually for default scheduler.

agrawalravi90 commented on src/server/run_sched.c:
```diff
@@ -465,10 +487,11 @@ schedule_jobs(pbs_sched *psched)
 		}
+	} else {
+		return 1;
+	}
```
Here and other places, please don't wrap a single line if or else block with curly braces

agrawalravi90 commented on src/server/run_sched.c:
```diff
@@ -492,53 +515,30 @@ schedule_jobs(pbs_sched *psched)
 	}
 
-	set_sched_sock(-1, psched);
+	psched->scheduler_sock = -1;
```
Why not use set_sched_sock() ?

suresh-thelkar replied:
I have removed this function but forgot to remove the header.  I cleaned it up now. Please check my latest commit.

agrawalravi90 commented on src/server/run_sched.c:
```diff
@@ -623,3 +623,128 @@ set_scheduler_flag(int flag, pbs_sched *psched)
+	for (psched = (pbs_sched*) GET_NEXT(svr_allscheds); psched; psched = (pbs_sched*) GET_NEXT(psched->sc_link)) {
+		if (psched->scheduler_sock2 == sock) {
+			DIS_tcp_funcs();
+			psched->sched_cycle_started = disrsi(sock, &rc);
```
psched->sched_cycle_started should be set to 0 regardless of what we receive on the channel right?

suresh-thelkar replied:
Yes. But here we are making sure that we indeed get 0 which acts as indication to say that sched cycle is finished.

agrawalravi90 commented on src/server/run_sched.c:
```diff
@@ -623,3 +623,128 @@ set_scheduler_flag(int flag, pbs_sched *psched)
+			scheduler_jobs_stat = 0;
+			handle_deferred_cycle_close();
+
+			if (rc == DIS_EOF || rc == -1)
```
The || rc == -1 is unnecessary

agrawalravi90 replied:
Also, how about setting rc = 0 at the top of the function, then you can just do return rc; at the end of the function

suresh-thelkar replied:
> The || rc == -1 is unnecessary

rc = -1 indicates an error. So we need it. Please let me know why do you that it is unnecessary.

suresh-thelkar replied:
> Also, how about setting rc = 0 at the top of the function, then you can just do return rc; at the end of the function

We cannot do this because as for as caller of this function is concerned it is expecting only two return values 0(success) and -1(error). But in the definition of the this function we call disrsi which can give either 0 or -1 or DIS_EOF. So inside this function we treat DIS_EOF and -1 as same and return -1.

suresh-thelkar replied:
Yes this check is not needed and removed it now. Please check my latest commit.

agrawalravi90 commented on src/server/sched_func.c:
```diff
@@ -619,7 +619,7 @@ action_sched_priv(attribute *pattr, void *pobj, int actmode)
 		return PBSE_SCHED_OP_NOT_PERMITTED;
 
-	if (actmode == ATR_ACTION_NEW || actmode == ATR_ACTION_ALTER || actmode == ATR_ACTION_RECOV) {
+	if (actmode == ATR_ACTION_NEW || actmode == ATR_ACTION_ALTER || actmode != ATR_ACTION_RECOV) {
```
why did you mke this change? We don't want this block to be executed if actmode is ATR_ACTION_NOOP or ATR_ACTION_FREE right?

suresh-thelkar replied:
The idea is not to do any validations which are happening in these action functions during ATR_ACTION_RECOV mode. This is because we store into database these validations are already carried out. There was some logical error in the above code and I have corrected it. Please take a look at my latest commit.

agrawalravi90 replied:
Then this should just be:
if (actmode == ATR_ACTION_NEW || actmode == ATR_ACTION_ALTER) right? by adding the || actmode != ATR_ACTION_RECOV, this validation code will also execute for ATR_ACTION_NOOP or ATR_ACTION_FREE, which is probably undesirable

suresh-thelkar replied:
I did exactly that i.e. removing act mode != ATR_ACTION_RECOV from the if conditions. Hope you are referring to my latest commit only or else I might be missing something. Please check this once again and let me know if you still feel I need to change anything.

agrawalravi90 replied:
hey, sorry, yes, this was comment on your old code, i saw the new code just now, thanks for clarifying

suresh-thelkar commented:
> Suresh there are again 144 files in the PR. Looks like many are from the rebase, can you instead cherry pick just ur changes to a new branch?

Sure. It looks like integration branch is again rebased with master yesterday and hence the issue. I am rebasing with integration branch and also working on an issue of jobs getting stuck intermittently after doing one more refactoring of combining primary and secondary scheduling fds as one chunk. I will fix this and let you know shortly.

suresh-thelkar commented:
> > Suresh there are again 144 files in the PR. Looks like many are from the rebase, can you instead cherry pick just ur changes to a new branch?
> 
> Sure. It looks like integration branch is again rebased with master yesterday and hence the issue. I am rebasing with integration branch and also working on an issue of jobs getting stuck intermittently after doing one more refactoring of combining primary and secondary scheduling fds as one chunk. I will fix this and let you know shortly.

@subhasisb, I have rebased my branch with the integration branch. It is open for further reviews. Requesting you to take a look and provide your review comments.

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -1662,6 +1662,10 @@ main(int argc, char **argv)
 		*pc = '\0';
 
+	if (server_init_type != RECOV_CREATE) {
+		connect_to_scheduler(dflt_scheduler);
```
This is related to failover. We currently support failover for only the default scheduler and hence we are connecting only to it at this point.

agrawalravi90 replied:
Can you please add a comment here explaining the same?

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -1684,13 +1688,13 @@ main(int argc, char **argv)
-	 * and fairshare usage.
-	 */
-	(void)contact_sched(SCH_CONFIGURE, NULL, pbs_scheduler_addr, pbs_scheduler_port);
-	(void)contact_sched(SCH_SCHEDULE_NULL, NULL, pbs_scheduler_addr, pbs_scheduler_port);
```
This was used to open up a secondary connection to the Scheduler. Here it is not required as secondary connection is already established by the time the control reaches here. Earlier we called connect_to_scheduler which establishes both primary and secondary connections.

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -1593,7 +1593,7 @@ main(int argc, char **argv)
 		if (server.sv_attr[(int)SRV_ATR_scheduling].at_val.at_long) {
 			/* Bring up scheduler here */
 			pbs_scheduler_addr = get_hostaddr(pbs_conf.pbs_secondary);
```
Yes. This is needed which is used to get the right host add in case of failover.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -267,66 +269,96 @@ find_sched_from_sock(int sock)
+	if (((which_conn == PRIMARY_CONN )&& (psched->scheduler_sock == -1) )||
+			((which_conn == SECONDARY_CONN) && (psched->scheduler_sock2 == -1))) {
+		/* Under win32, this function does a timeout wait on the non-blocking socket */
+		sock = client_to_svr(psched->pbs_scheduler_addr, psched->pbs_scheduler_port, 1); /* scheduler connection still uses resv-ports */
```
> I think we should rename this function (client_to_svr) to something like (connect_to_host) since here the server is using it to connect to scheduler.

I haven't renamed this function to client_to_svr. Moreover this is getting called from multiple places. I feel better not to rename this in this PR.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -267,66 +269,96 @@ find_sched_from_sock(int sock)
+		conn->cn_authen |= PBS_NET_CONN_TO_SCHED | PBS_NET_CONN_FROM_PRIVIL | PBS_NET_CONN_AUTHENTICATED;
+
+		if (which_conn == SECONDARY_CONN)
+			conn->is_sched_conn = 1;
```
I have now renamed the variable to conn->is_sched_second_conn. Since Server receives end of cycle notification only on secondary connection we are interested only in it.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -267,66 +269,96 @@ find_sched_from_sock(int sock)
-		snprintf(log_buffer, sizeof(log_buffer), "cannot set nodelay on connection %d (errno=%d)\n", sock, errno);
-		log_err(-1, __func__, log_buffer);
-		return (-1);
+	if (sock == -1) {
```
This was required as per the earlier code. Since I have refactored this code, now it is not needed.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -267,66 +269,96 @@ find_sched_from_sock(int sock)
-		log_err(-1, __func__, log_buffer);
-		return (-1);
+	if (sock == -1) {
+		if ((which_conn == PRIMARY_CONN) && (psched->scheduler_sock != -1))
```
Thanks for suggesting this. I would like to do something similar in the next PR, but sine I got this review comment, I have implemented this in the current PR itself. Instead of declaring the array with scheduler_sock[FINAL_CONN] I declared it as scheduler_sock[PRIMARY_CONN+1] since it is more readable to say we need two connections. Other than this everything is implemented as described above. Please look into my latest commit for changes and let me know if there needs any further changes.

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -1745,8 +1752,8 @@ main(int argc, char **argv)
-					if ((psched->scheduler_sock2 != -1) &&
-						(psched->scheduler_sock != -1)) {
+					if ((psched->scheduler_sock2 != -1) && (psched->scheduler_sock != -1) &&
+							psched->sched_cycle_started == 1) {
```
I believe you are talking about line 1773 instead of 1173.
"psched->sch_attr[(int)SCHED_ATR_scheduling].at_val.at_long" is used to know whether scheduling is enabled or not. We cannot tell whether sched cycle is ongoing or not with this. 
If we want to know whether sched cycle is in progress then we need to check sched_cycle_started flag to 1.

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -1849,7 +1856,7 @@ main(int argc, char **argv)
 
 	if (brought_up_alt_sched == 1)
-		(void)contact_sched(SCH_QUIT, NULL, pbs_scheduler_addr, pbs_scheduler_port);
+		(void)contact_sched(SCH_QUIT, NULL,  dflt_scheduler, PRIMARY_CONN);
```
SCH_QUIT is sent on primary connection only even before my change. Only command that is sent on secondary connection is SCH_SCHEDULE_RESTART_CYCLE. We have to change the existing design if we want to send it over secondary_connection. Just curious to know why are you thinking to change it to secondary connection and is there any advantage with it ?

agrawalravi90 replied:
Oh, interesting .. SCH_QUIT sounded like a super high priority command to me, it'll make the scheduler quit immediately instead of possibly sending job run requests to server when the server might be on its way to dying. But if that's how it was before then it's up to you whether you want to change this behavior or not.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -358,23 +392,17 @@ schedule_high(pbs_sched *psched)
-		if ((s = contact_sched(psched->svr_do_sched_high, NULL, psched->pbs_scheduler_addr, psched->pbs_scheduler_port)) < 0) {
-			set_attr_svr(&(psched->sch_attr[(int) SCHED_ATR_sched_state]), &sched_attr_def[(int) SCHED_ATR_sched_state], SC_DOWN);
+	if (psched->sched_cycle_started == 0) {
+		if ((s = contact_sched(psched->svr_do_sched_high, NULL, psched, PRIMARY_CONN)) < 0) {
```
As I have explained earlier only super high priority commands like SCH_SCHEDULE_RESTART_CYCLE are sent on secondary connection. Rest all are sent on primary connection only.

agrawalravi90 commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -455,34 +483,68 @@ int
+	}
+
+	/* If ch_shards table is filled up then we can derive inact_svr_indexes */
+	if (ch_shards_filled) {
```
if ch_shards_filled was not 1 then you'd not reach this line of code, so this if condition is not needed.

suresh-thelkar replied:
This is now needed as we renamed the function to get_conn_shards_filled which gets this value of the flag ch_shards_filled. Its value can be either 0 or 1 or -1. We want to execute the code following the if block only if its value is 1 and hence needed. Please look into my latest commit.
Also if we don't have this if block then this code will also be executed by the normal clients like qsub, qstat etc. and MOM which we don't want.

suresh-thelkar replied:
Yes. For your information I have refactored the above code as per Subhasis's comments. Please take a look at my latest commit for more details.

agrawalravi90 commented on src/server/run_sched.c:
```diff
@@ -267,78 +267,96 @@ find_sched_from_sock(int sock)
 
-	/* connect to the Scheduler */
-	/* put a timeout alarm around the connect */
+	if (psched->scheduler_sock[which_conn] == -1) {
```
Do you wanna check here that which_conn is less than 2? This is why I had suggested adding another constant FINAL_CONN to to sched_conn_t so that you could check here that "which_conn < FINAL_CONN".

suresh-thelkar replied:
No. We don't have a need to check whether which_conn < 2. All we need to know here is whether the given connection is established or not. If not i.e. if (psched->scheduler_sock[which_conn] == -1) then we establish a new connection else we continue with the existing connection.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1472,232 @@ main(int argc, char *argv[])
+		shard_conn_t **shard_conn;
+
+		/* Use virtual socket i.e. server_sock when calling get_conn_shards */
+		shard_conn = (shard_conn_t **)get_conn_shards(server_sock);
```
Yes server_sock is actually a real socket which is used for listening to incoming connections. We are also using this as virtual socket in case of Multiple Servers only. This is inline with the Libshard design and its integration with IFL.
Modified the comment to make it more clear on this. Please look into my latest commit.

suresh-thelkar commented on src/include/pbs_sched.h:
```diff
@@ -66,6 +66,15 @@ extern "C" {
+/* Maximum server index possible
+ * This means we can have maximum 99 servers with index 0 to 98.
+ */
+#define SVR_MAX_INDEX 2
```
I think it is not good to have hardcoded numbers inside our code and hence used a macro. I have renamed the macro to SVR_MAX_INDEX_STR_LEN. I feel this is more clear now. Also added a very detailed comment describing the purpose of it. Please check my latest commit and let me know what do you think.

agrawalravi90 replied:
Is this used in more than 1 function? If not then I feel that this is overkill, how about just adding a const int buflen inside the function itself?

suresh-thelkar replied:
No. For now it is used only in one function. I have now added a const int SVR_MAX_INDEX_STR_LEN inside the function itself along with the detailed comment to describe the purpose of it. Please take a look at my latest commit.

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -662,3 +662,74 @@ get_conn_shards(int vfd)
+
+/**
+ * @brief
+ * 	get_conn_shards_exist - Fetches ch_shards_exist through which we know
```
I now think that we need to name this function as get_conn_shards_filled(). This is because it actually fetches the value of the flag ch_shards_filled. I have added the comment accordingly.

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -662,3 +662,74 @@ get_conn_shards(int vfd)
+ * @param[in] vfd - Virtual socket number
+ *
+ * @return int
+ * @retval  0 - success 
```
As I said in the earlier comment I have now renamed this function to get_conn_shards_filled() which can actually returns the value of the flag ch_shards_filled which can be 0 or 1. It can also return -1 in case of failure. Also modified the header accordingly. Please check my latest commit.

hirenvadalia commented on src/scheduler/pbs_sched.c:
```diff
@@ -284,6 +292,70 @@ server_disconnect(int connect)
+			if (shard_connection[index_to_shards]->secondary_sd >= 0) {
+				close_tcp_connection(shard_connection[index_to_shards]->secondary_sd);
+				/* unlock the connection level lock */
+				pbs_client_thread_unlock_connection(server_sock);
```
I think you have to put this after this if and below if, else you will be unlocking multiple time,

Or better take out locking and unlocking outside of `if (get_max_servers() > 1) .. else ..` block

suresh-thelkar commented on src/include/libpbs.h:
```diff
@@ -144,6 +145,8 @@ int set_conn_chan(int, pbs_tcp_chan_t *);
 pthread_mutex_t * get_conn_mutex(int);
 int set_conn_shard_context(int, int);
 int get_conn_shard_context(int);
+int set_conn_shards_filled(int);
```
@subhasisb, As per our discussion I have come up with a dummy function internal_connect_sched which is called via pfn_connect function pointer which allowed me to eliminate the above accessors or mutators. Thanks for suggesting this method. Please look into my latest commit for more details and provide your further comments.

hirenvadalia replied:
I thought you will be removing these methods? no?

suresh-thelkar replied:
Yes. These methods are no longer needed and I did the clean up. Please look into the latest commit.

suresh-thelkar commented on src/include/net_connect.h:
```diff
@@ -221,6 +221,7 @@ struct connection {
 	void            *cn_data;         /* pointer to some data for cn_func */
 	char            cn_username[PBS_MAXUSER + 1];
 	char            cn_hostname[PBS_MAXHOSTNAME + 1];
+	int		is_sched_second_conn; /* used to know whether this is Scheduler's secondary connection or not */
```
I have now named it conn_origin which is actually an enum as
```
/* This is used to know where the connection is originiated from.
 * This can be extended to have MOM and other clients of Server in future.
 */
typedef enum conn_origin {
	SCHED_PRIMARY,
	SCHED_SECONDARY
} conn_origin_t;
```
Please look into my latest commit for more details.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -436,6 +438,32 @@ int internal_client_connect(int vsock, char *server, int port, char *extend_data
+ * @retval	NULL if fails
+ */
+static int *
+initialise_inact_svr_indexes()
```
This function is getting used only in get_svr_shard_connection(). MOM and Scheduler are not using it.

suresh-thelkar commented:
@subhasisb / @agrawalravi90 / @hirenvadalia  , I have addressed your comments and did some more miscellaneous changes. Please take a look at my latest commit.

hirenvadalia commented on src/include/pbs_sched.h:
```diff
@@ -103,17 +107,26 @@ enum sched_atr {
+/* This is used when connecting to the Scheduler to let Server know on which
+ * connection(either primary or secondary) it has to talk to.
+ */
+enum sched_conn_t {
```
Can't we use conn_origin_t instead this?

suresh-thelkar replied:
No. We cannot use the conn_origin_t here which is used to know the origin of the connection i.e. whether the connection is from scheduler, MOM, clients etc. where as sched_conn_t is specific to scheduler connections.

agrawalravi90 replied:
I kind of agree with Hiren here, why not just use conn_origin_t here? conn_origin_t can also tell the server that the connection was primary or secondary sched, so this seems redundant.

suresh-thelkar replied:
The reason why I did not use the same enum initially for both is that one enum(conn_origin_t) talks about from where the connection is originating from like MOM,SCHED etc. and the other enum talks about whether scheduler connection(primary/secondary) is established or not. I felt these two are different things even though it might seem little redundant.

Anyways I am now using the same enum i.e. conn_origin_t and deleted the other enum and all references to it.

hirenvadalia commented on src/lib/Libnet/net_server.c:
```diff
@@ -772,6 +772,7 @@ add_conn_priority(int sd, enum conn_type type, pbs_net_t addr, unsigned int port
 	conn->cn_authen = 0;
 	conn->cn_prio_flag = 0;
 	conn->cn_auth_config = NULL;
+	conn->conn_origin = -1;
```
Define SCHED_UNDEFINED in conn_origin_t and use that here instead -1

suresh-thelkar replied:
Not required. We cannot use something like SCHED_UNDEFINIED for the above. It is because at least one scheduler is always there and defined. 
I used -1 instead of 0 because 0 is valid for conn_origin since it is a enum.

hirenvadalia replied:
How about SCHED_UNKNOWN? My point is `conn->conn_origin` should come from conn_origin_t enum right? and here -1 is something hardcoded, what if someone wants to check this against whether it has valid value or not? (like `if (conn->conn_origin != SCHED_UNKNOWN)`) then as per current code, you have to like `if (conn->conn_origin != -1)` which is error prone.

suresh-thelkar replied:
We can use SCHED_UNKNOWN. But just eager to know what is the use case behind using `if (conn->conn_origin != SCHED_UNKNOWN)`
All we need to check is something like `if (conn->conn_origin == SCHED_SECONDARY) {` or `if (conn->conn_origin == SCHED_PRIMARY) {`. Both of these solves our purpose. If there is such a need we can do it in future. conn_origin is open for extending and it is generic. Tomorrow we can add MOM, other clients in the above enum during which case if we again add SCHED_UNKNOWN it becomes very much specific to scheduler which is something that we don't want.

hirenvadalia replied:
Yes use case is it can be a replacement for `PBS_NET_CONN_TO_SCHED` in `conn->cn_authen` (maybe not now but in future, because  as of now `PBS_NET_CONN_TO_SCHED` used to indicate the connection is to/from sched only, it has nothing to do with connection authentication)

suresh-thelkar replied:
Thanks @hirenvadalia for sharing the above. If we are planning to replace PBS_NET_CONN_TO_SCHED with conn_origin in future then we should refactor code accordingly and I feel this is outside of the scope of the current PR. As said earlier this enum is open for extensions and we always can enhance it as per our future needs.

hirenvadalia replied:
Well, @suresh-thelkar, I feel it's well within this PR's scope because you added conn->conn_origin to know whether the connection is from sched or not, and if it's from sched then which connection like primary or secondary, so ultimately you are checking for `is the connection to/from sched?` and as of now `PBS_NET_CONN_TO_SCHED` is used only to for that purpose only, so in my view (and I strongly feel we should do it because on other hands we are trying to clean PBS code also, and this is one of that chance to do it, so why not do it?), we should remove `PBS_NET_CONN_TO_SCHED` in this PR only. But if you still feel that it's not then I will leave that up to you and @subhasisb to decide.

suresh-thelkar replied:
We can refactor our code to eliminate PBS_NET_CONN_TO_SCHED and we can take it in our next enhancements. That way it is good for tracking/bugs finding/fixing. We also need to test authentication code also if we club this with the current PR. Already there are many things clubbed in the current PR. In future if we get a bug it is really difficult to track if we club so many things in the same PR. As you are leaving it to me and Subhasis, I personally feel it would be better if we do it later. I would also like to hear from Subhasis and proceed further accordingly.

subhasisb replied:
If the change is small we can do it in this PR itself. It will also justify the introduction of that enum better

hirenvadalia replied:
Yes change is like 5-6 line change in total (but 1 line change at 4-5 different place)

subhasisb replied:
about SCHED_UNKNOWN, better to call it just UNKNOWN and add it to the enum as -1

hirenvadalia commented on src/scheduler/fifo.c:
```diff
@@ -2525,8 +2526,7 @@ sched_settings_frm_svr(struct batch_status *status)
-				err = pbs_manager(connector,
-					MGR_CMD_SET, MGR_OBJ_SCHED,
+				err = pbs_manager(sd, MGR_CMD_SET, MGR_OBJ_SCHED,
 					sc_name, attribs, NULL);
```
make arguments in one line, here and other places too

suresh-thelkar replied:
I have put all the arguments in the above and similar places wherever it is reasonable. 

But there are some places we cannot put everything in one line which hurts readability. And also I see at many places in PBS code either the function signature or function call is spanned across multiple lines for better readability.  Just curious to know if we have a coding standard for this. If yes can you please point me to that.

hirenvadalia commented on src/scheduler/get_4byte.c:
```diff
@@ -66,8 +66,9 @@
- * @param[in]	jid	-	if 'val' obtained is SCH_SCHEDULE_AJOB, then '*jid'
- *						holds the jobid of the job to be scheduled.
+ * @param[in]	identifier	-	if 'val' obtained is SCH_SCHEDULE_AJOB, then '*identifier" is  job id.
+ *					holds the jobid of the job to be scheduled.
```
`then '*identifier" is the job id of the job to be scheduled.`

hirenvadalia commented on src/scheduler/globals.h:
```diff
@@ -45,6 +45,12 @@ extern "C" {
 #include "queue.h"
+
+typedef struct pbs_svr_sock_pair {
+	int ch_socket;
```
can we rename this to like `ch_primary_socket`?

hirenvadalia commented on src/scheduler/pbs_sched.c:
```diff
@@ -1350,59 +1428,42 @@ main(int argc, char *argv[])
-				update_svr = 0;
-			}
+		if (schedule_wrapper(max_svrs, num_cfg_svrs, &update_svr,
+			&read_fdset, opt_no_restart) == 1)
```
make it one line please

hirenvadalia commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1474,255 @@ main(int argc, char *argv[])
+ */
+static int
+schedule_wrapper(int max_svrs, int num_cfg_svrs, int *update_svr,
+		fd_set *read_fdset, int opt_no_restart)
```
one line please

hirenvadalia commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1474,255 @@ main(int argc, char *argv[])
+			if (ret != 1) {
+				close_server_conn(shard_inst_idx);
+				log_eventf(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SCHED, LOG_ERR, __func__,
+					"get_sched_cmd failed, errno=%d in function %s. "
```
add one more tab to align propely

hirenvadalia commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1474,255 @@ main(int argc, char *argv[])
+ * @brief
+ *      Dummy function used for Scheduler to connect Server
+ *
+ *	As of today Scheduler does not connect o Server but Server only connects to Scheduler
```
`connect **to** server`

hirenvadalia commented on src/scheduler/pbs_sched.c:
```diff
@@ -1413,3 +1474,255 @@ main(int argc, char *argv[])
+ *
+ *	As of today Scheduler does not connect o Server but Server only connects to Scheduler
+ *	and hence this is a dummy function which just returns -1 always. 
+ *      For more details how this function is getting used, please refer to the function
```
indentation fix please

suresh-thelkar replied:
I think it is GitHub issue. I don't see any issue in formatting when I opened the file in vi editor.

hirenvadalia replied:
It's not GitHub issue, its because you have mix of space and tab,
![Screen Shot 2020-05-22 at 2 06 48 PM](https://user-images.githubusercontent.com/17191391/82648673-a396f300-9c35-11ea-8106-711d40a0834e.png)

suresh-thelkar replied:
Fixed the tabs in the function getter. Please check latest commit.

hirenvadalia commented on src/server/pbsd_main.c:
```diff
@@ -1662,6 +1662,10 @@ main(int argc, char **argv)
 		*pc = '\0';
 
+	if (server_init_type != RECOV_CREATE)
+		/* Since we support failover only for default scheduler we connect only to it  at this place */
```
extra space between `only to it` and `at this`

hirenvadalia commented on src/server/pbsd_main.c:
```diff
@@ -1734,7 +1738,11 @@ main(int argc, char **argv)
 				/* if time or event says to run scheduler, do it */
-
+				if (psched->scheduler_sock[PRIMARY_CONN] == -1 &&
+					psched->scheduler_sock[SECONDARY_CONN] == -1) {
```
one line please, please to all your changes

suresh-thelkar replied:
Same as my earlier comment.

hirenvadalia commented on src/server/run_sched.c:
```diff
@@ -267,78 +267,102 @@ find_sched_from_sock(int sock)
+	 * As we are restricting number of servers to 99 in the worst case(although in reality one can use at the max 5 to 10 servers),
+	 * if we convert these ids to string form, we hardly need a char array of length two + 1 for NULL character.
+	 */
+	const int SVR_MAX_INDEX_STR_LEN = 3;
```
shouldn't this be macro?

suresh-thelkar replied:
As per the reviewer's suggestion I am not using a macro or constant and using the hardcoded number 3 with the very detailed comment describing its purpose.

suresh-thelkar commented:
@hirenvadalia, I have addressed your comments. Please take look at the latest commits.

@subhasisb / @agrawalravi90, Can you please go through my recent commits and provide further review comments if any.

subhasisb commented on src/include/pbs_sched.h:
```diff
@@ -103,17 +107,26 @@ enum sched_atr {
 	pbs_list_link	sc_link;		/* forward/backward links */
-	int scheduler_sock;
-	int scheduler_sock2;
+	int scheduler_sock[PRIMARY_CONN + 1];
```
why not just say 2?

suresh-thelkar replied:
Just to avoid the hard coded number I did not use 2 directly.
Now I have changed the code as you suggested.

agrawalravi90 commented:
Also, please post performance numbers from running the sched performance test, thanks.

suresh-thelkar commented:
> Suresh, I don't really have further comments, other than cosmetics, which can be discussed in the OSS PR. Can you please post TH3 results here (with extensions)?

Thanks for all your reviews Ravi. I have run TH3 several times. Attached is the recent snapshot. 
<img width="451" alt="image" src="https://user-images.githubusercontent.com/17923736/82745699-f216cf80-9da4-11ea-92c3-0409d0ff5376.png">

Note: There are no new failures. Test cases that are failed are known issues and exists in our integration branch.

I am also planning to run TH3 once again after the final sign off just to make sure that everything is fine.

suresh-thelkar commented:
> Also, please post performance numbers from running the sched performance test, thanks.

As we discussed in order to compare performance results before and after my change, I have first started running these tests in integration branch. Only 1 out of 5 test cases passed and the complete report was also not generated.  
Except the persistent connections all other things done in this PR are not directly related to performance, for now I will make a note to run all the performance test cases as soon as they are fixed in integration branch. Hoping that this works for all of us.

suresh-thelkar commented:
> close just a few more things

Thanks for all your reviews Subhasis. I am the process of implementing and testing the final comments you suggested. I will let you know once done.

suresh-thelkar commented:
Following are the changes made since last commit.
- Refactoring the code to eliminate PBS_NET_CONN_TO_SCHED
- Addressed the comments like to have UNKNOWN=-1 as a member of conn_origin_t  and use this enum itself instead of sched_conn_t.

@subhasisb/ @hirenvadalia/ @agrawalravi90, Please go through the latest commit and let me know If I have missed anything.
Also all the UNIT testing for single server and the multi server path to the extent that we can test are successful. Attaching the logs for the same.
[UnitTesting.log](https://github.com/subhasisb/openpbs/files/4675292/UnitTesting.log)

suresh-thelkar commented:
@hirenvadalia, Thanks for your review and sign off.

subhasisb commented on src/server/pbsd_main.c:
```diff
@@ -1738,8 +1738,8 @@ main(int argc, char **argv)
 				/* if time or event says to run scheduler, do it */
-				if (psched->scheduler_sock[PRIMARY_CONN] == -1 &&
-					psched->scheduler_sock[SECONDARY_CONN] == -1) {
+				if (psched->scheduler_sock[SCHED_PRIMARY] == UNKNOWN &&
```
better not to use the enum members as indexes

subhasisb replied:
also using them both as index and value is overloading the meaning

suresh-thelkar replied:
Thanks for your suggestion. I have modified the code accordingly. Please look into my latest commit.

suresh-thelkar commented:
@subhasisb, Thanks very much for your thorough review and sign off.  Just adding the following information for the benefit of others.

Changes made since last commit are
- Implemented Subhasis's review comments
- Change to fix windows build issue(Actually there was a clash for the enum member UNKNOWN with a specific enum for Windows only and hence renamed conn_origin_t members and made other things as suggested by Subhasis)

subhasisb commented:
since @agrawalravi90 had also done a thorough review, we will wait for his comments or sign off before merging

agrawalravi90 commented:
Looks good to me, I just think that you should run another TH3 regression before we merge. Thanks.

subhasisb commented:
also now conflicts needs to be resolved

suresh-thelkar commented:
> Looks good to me, I just think that you should run another TH3 regression before we merge. Thanks.

Thanks @agrawalravi90 for your thorough review and sign off.

suresh-thelkar commented:
> also now conflicts needs to be resolved

@subhasisb, I have resolved all the conflicts and pushed my changes.

suresh-thelkar commented:
@subhasisb, I have squashed, rebased and resolved all the conflicts. Have done one more round of testing after this. All the testing went fine. Can you please take this PR forward. Following are the details of testing.

Attached is the UNIT testing logs for MultiServer code path to the extent that we can test. All test cases are passing.

1. [UnitTesting.log](https://github.com/subhasisb/openpbs/files/4680538/UnitTesting.log)


2.Given below is the snapshot of TH3. Except the known issues all regression is good.
![image](https://user-images.githubusercontent.com/17923736/82873057-8faf0200-9f51-11ea-873a-42d09c030277.png)
